distributed/scheduler.py

```diff
@@ -2328,6 +2335,44 @@ def decide_worker(self, ts: TaskState) -> WorkerState:
             ts.state = "no-worker"
             return ws
 
+        total_nthreads: Py_ssize_t
+        if valid_workers is None:
+            total_nthreads = self._total_nthreads
+        else:
+            total_nthreads = 0
+            for ws in valid_workers:
+                total_nthreads += ws._nthreads
+
+        group_tasks_per_thread: double = (
+            (len(group) / total_nthreads) if total_nthreads > 0 else 0
+        )
+        if group_tasks_per_thread > 2 and sum(map(len, group._dependencies)) < 5:
+            # Group is larger than cluster with very few dependencies; minimize future data transfers.
+            ws = group._last_worker
+            if not (ws and valid_workers is not None and ws not in valid_workers):
+                if (
+                    ws
+                    and ws._occupancy / ws._nthreads / self.get_task_duration(ts)
+                    < group_tasks_per_thread
+                ):
+                    # Schedule sequential tasks onto the same worker until it's filled up.
+                    # Assumes `decide_worker` is being called in priority order.
+                    return ws
+
+                # Pick a new worker for the next few tasks, considering all possible workers
+                worker_pool = (
+                    valid_workers
+                    if valid_workers is not None
+                    else (self._idle_dv or self._workers_dv).values()
+                )
+                ws = min(
+                    worker_pool,
+                    key=partial(self.worker_objective, ts),
+                    default=None,
+                )
+                group._last_worker = ws
+                return ws
+
```

Comment:
I've only gone through this with a shallow pass, but at first glance it seems to make sense to me.

It was also a lot easier to go through for me.  I appreciate you accomodating my lack of attention span here :)

Reply:
I prefer this too.

The main thing was doing enough tests to convince myself that being called out of priority order was not a case we needed to worry about. I thought it happened during a shuffle, but actually it was only happening while the cluster was shutting down at the end of the script after doing a shuffle, and the scheduler was trying to rerun tasks from workers that were leaving. And even if we do get called out of priority order, the occupancy metric will keep us from oversaturating any one worker, and if we happen to re-assign to a root-ish task to a worker where it doesn't have many relatives, it's still probably no worse than the typical `decide_workers` path.